src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinResultEdgeReader.java

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.spring.data.gremlin.conversion;
+
+import com.microsoft.spring.data.gremlin.common.Constants;
+import com.microsoft.spring.data.gremlin.exception.UnexpectedGremlinSourceTypeException;
+import org.apache.tinkerpop.gremlin.driver.Result;
+import org.springframework.lang.NonNull;
+import org.springframework.util.Assert;
+
+import java.util.Map;
+
+public class GremlinResultEdgeReader extends BasicGremlinResultReader implements GremlinResultReader {
+
+    public GremlinResultEdgeReader() {
+        super();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void read(@NonNull Result result, @NonNull GremlinSource source) {
+        if (!(source instanceof GremlinSourceEdge)) {
+            throw new UnexpectedGremlinSourceTypeException("Should be instance of GremlinSourceEdge");
+        }
+
+        Assert.isInstanceOf(Map.class, result.getObject(), "should be one instance of Map");
+        final Map<String, Object> map = (Map<String, Object>) result.getObject();
+
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_ID), "should contain id property");
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_LABEL), "should contain label property");
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_TYPE), "should contain type property");
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_PROPERTIES), "should contain properties property");
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_INV), "should contain inV property");
+        Assert.isTrue(map.containsKey(Constants.PROPERTY_OUTV), "should contain outV property");
+        Assert.isTrue(map.get(Constants.PROPERTY_TYPE).equals(Constants.RESULT_TYPE_EDGE), "must be vertex type");
```

Comment:
I don't know whether it's good or not to use assertion here to check keys exist, as the assertions will throw Runtime exceptions, which means user may not handle possible assertion failures(at least it's not normal practice to handle RuntimeException), and the program will totally exit once assertion fails. User cannot catch the exception and recover from such failures.

Reply:
https://stackoverflow.com/questions/2758224/what-does-the-java-assert-keyword-do-and-when-should-it-be-used#comment27084614_2758645

https://softwareengineering.stackexchange.com/a/137163

I think in this repo we are using a lot of `Assert`s, could you share the reason why we are using so many Assertions?

Can we switch to `Exception`s for most `Assert`s?

Reply:
https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html

**Do not use assertions for argument checking in public methods.**

**Do not use assertions to do any work that your application requires for correct operation.**
`Because assertions may be disabled,...`

Reply:
OK, sorry just missed you are using Assert in Spring framework, still it's Runtime exception and user will not recover without catching a Runtime Exception.

Reply:
In my opinion, assert is for something that it will not happen (unexpected) in normal env (most situations). And exception is for something that we may expected like user invalid input. Assert is designed to be ignored in release version, and for internal assurance which is not suitable for external inputs/data.

Here I use exception for argument checking, and assert for what it should contains. Because most cases, it come from gremlin server, and should be stable in short future. So we don't need to check that, and assert is enough here.